File: python/src/byoc/providers/gdrive/_pkce.py
"""PKCE (RFC 7636) helpers for the Google Drive OAuth flow.

PKCE is what stops an intercepted authorization code from being redeemed by an
attacker, so the verifier must come from a cryptographically secure RNG -- never
``random``. Behaviour is pinned by ``spec/fixtures/pkce.json``, whose first
vector is RFC 7636 Appendix B.
"""

from __future__ import annotations

import base64
import hashlib
import secrets

from ...errors import InvalidInputError

MIN_VERIFIER_LENGTH = 43
MAX_VERIFIER_LENGTH = 128

# RFC 7636 section 4.1: verifiers use the unreserved character set.
_VERIFIER_ALPHABET = (
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~"
)


def _base64url(raw: bytes) -> str:
    """Base64url-encode without padding, per RFC 7636."""
    return base64.urlsafe_b64encode(raw).decode("ascii").rstrip("=")


def generate_code_verifier(length: int = 64) -> str:
    """Generate a cryptographically random ``code_verifier``.

    Args:
        length: Desired length, clamped to the RFC's 43..128 range.
    """
    size = max(MIN_VERIFIER_LENGTH, min(MAX_VERIFIER_LENGTH, length))
    return "".join(secrets.choice(_VERIFIER_ALPHABET) for _ in range(size))


def generate_code_challenge(verifier: str) -> str:
    """Derive the S256 ``code_challenge`` from a verifier.

    The challenge is the base64url-encoded SHA-256 of the ASCII verifier.
    """
    if not verifier:
        raise InvalidInputError("PKCE code_verifier must not be empty.", provider="google-drive")
    return _base64url(hashlib.sha256(verifier.encode("ascii")).digest())


def generate_oauth_state(length: int = 32) -> str:
    """Generate a random ``state`` value for CSRF protection.

    The caller must store this and compare it against the value returned on the
    redirect; skipping that check leaves the flow open to CSRF.
    """
    return secrets.token_urlsafe(max(16, length))[:length]

File: packages/google-drive/src/auth/pkce.ts
import { webcrypto } from "node:crypto";

const cryptoObj = (typeof crypto !== "undefined" ? crypto : webcrypto) as Crypto;

/**
 * Converts an ArrayBuffer to a base64url-encoded string (RFC 7636).
 */
function base64UrlEncode(buffer: ArrayBuffer): string {
  const bytes = new Uint8Array(buffer);
  let binary = "";
  for (let i = 0; i < bytes.byteLength; i++) {
    binary += String.fromCharCode(bytes[i]!);
  }
  return btoa(binary)
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/, "");
}

/**
 * Generates a cryptographically random code_verifier string (43 to 128 characters).
 */
export function generateCodeVerifier(length: number = 64): string {
  const validLength = Math.max(43, Math.min(128, length));
  const randomBytes = new Uint8Array(validLength);
  cryptoObj.getRandomValues(randomBytes);
  return base64UrlEncode(randomBytes.buffer).substring(0, validLength);
}

/**
 * Generates an S256 code_challenge from a code_verifier (RFC 7636).
 */
export async function generateCodeChallenge(verifier: string): Promise<string> {
  const encoder = new TextEncoder();
  const data = encoder.encode(verifier);
  const digest = await cryptoObj.subtle.digest("SHA-256", data);
  return base64UrlEncode(digest);
}

/**
 * Generates a cryptographically secure random state string to protect against CSRF.
 */
export function generateOAuthState(length: number = 32): string {
  const randomBytes = new Uint8Array(length);
  cryptoObj.getRandomValues(randomBytes);
  return base64UrlEncode(randomBytes.buffer).substring(0, length);
}
